Bugfix - Add FilePathNioAsync to solve the poor direct buffer allocation issue and the mvstore closed issue by client thread interrupted in FileChannel

forchid opened the pull request:
Fix the issue #1502, and no performance loss(keeping 3,000,000~5,000,000 get-put operations per second).

forchid committed:
Bugfix - Write page by page(or in the unit of a few pages) for avoiding to allocate direct buffer poorly in the implementation "sun.nio.ch.FileChannelImpl" of java.nio.channels.FileChannel

forchid committed:
* "Write page by page" moved to FilePathNio

andreitokar commented:
IMHO, this is not an issue to fix in MVStore, see my comment in #1502. It has been intention to increase size of a single write operation to improve performance, especially in SSD case, therefore splitting it into smaller buffers does not make much sense. In extreme cases, like the one presented in #1502,  jdk.nio.maxCachedBufferSize can be used instead.

forchid commented:
The "-Djdk.nio.maxCachedBufferSize" only limits the cached buffer size, but can't limit the allocated max direct buffer size in FileChannel, and only can be used after JDK-1.8.0b102.

forchid commented:
@andreitokar The unsuitable size setting of "-Djdk.nio.maxCachedBufferSize" can lead to callocate direct buffer frequently, and performance loss. The suitable size setting of "-Djdk.nio.maxCachedBufferSize" is very hard for h2 MVStore WriteBuffer.

andreitokar commented:
Picking unsuitable value for any JDK setting can do a harm, but in this case it is not that hard to peek: 1 - 4MB would work, unless memory is tight. But again, we are talking about some extreme cases, which should be avoided in the first place. Otherwise, default will work just fine. Limiting buffer size by smaller value may degrade performance of SSD drive, but will conserve RAM.

forchid committed:
+ Add FilePathNioAsync to solve the poor direct buffer allocation issue
and the closed issue by application thread interrupted in java nio
FileChannel, and no regression

andreitokar commented:
Why all of the operations, and not just writes, should go through the thread pool? 
What is the reasoning behind having a separate "read" pool?
Operations in question are primarily I/O and not CPU bound, so why size of the pool is limited by CPU count? 
BTW, IDEA complains about missorted "final static" modifiers and possibility of "private static final".

forchid commented:
@andreitokar 
Answers1 and 2: 
  Protecting file store from closed by session thread interrupted when reading.
Answer 3: 
  More IO threads maybe lead to IO overhead. The size of read pool can be given bigger.

katzyn commented:
I do not understand your motivation. You built an artificial test case using a some `-XX` flags, found a bug in JDK, and trying to build a workaround for this bug without taking care of real-world use cases? Why?

I didn't test your changes yet, but I have a some questions.

1. Why `ExecutorService` instances are not closed anywhere? H2 is used as embedded database by many web applications and may be packaged together with an application due to different reasons. When such application is reloaded / upgraded / unloaded it must free all its resources to prevent a memory leak. With current H2 it's enough to flush all necessary data and shutdown the database from `ServletContextListener.contextDestroyed()`. But your new threads will stay alive. Tomcat, for example, reports a memory leak and tries to terminate such threads by itself and I don't see any way to avoid it with your implementation. The same issue may happen in other environments where libraries can be unloaded, but not all containers are smart enough to plug the leak somehow.

2. If database server is running (another use case) with multiple databases massive writes in some databases will block all other databases? If virtual server has only one or two dedicated CPU cores only one or two write operations may be performed in parallel? It may create a lot of problems. Each database should have a chance to read or write something in reasonable time.

3. JDK provides own [AsynchronousFileChannel](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/AsynchronousFileChannel.html) that works far better with scattered operations on some systems (Windows / NTFS and may be some others) than any other ways possible in Java without JNI. Of course, it cannot be used unconditionally because it is not available on Android before API 26, but it is available on all other platforms that we support. I don't think that reinvention of wheel with custom implementation is needed.

katzyn commented:
So from my point of view it's OK to use Java's own built-in asynchronous capabilities by default when they are available (with possible automatic fallback to old `FilePathNio` for Android), but it should be carefully tested, it also may raise some new issues in environments that are not covered by automated tests. If you want to try this way please note that `java.nio.file.Path` is also not available on older versions of Android.

Proposed custom implementation doesn't look like something reasonable.

forchid committed:
* Modify the life-cycle of read/write pool is the same of FileStore in
FilePathNioAsync.
* Keep interrupted status if current thread interruped in
FilePathNioAsync.

forchid commented:
@katzyn Thank you! My motivation in this PR is also that:
Protecting java.nio.channels.FileChannel or mvstore from closed by session thread interrupted.Please see the [issue 227](https://github.com/h2database/h2database/issues/227).

For your given comment, I has modified the life-cycle of the read/write pool in FilePathNioAsync. Now these pools only keep running in the life-cycle of FileStore.

Uncle-pan committed:
* fix - close RandomAccessFile issue in constructor.

katzyn commented:
There is a `FilePathRetryOnInterrupt` for a few people that use interrupts in a database application. This issue is not very significant.

> Now these pools only keep running in the life-cycle of FileStore.

It's only a waste of resources for a database server and you can easily got an original problem with excessive usage of native memory due to bugs in some JDKs.

Try to use native Java implementation, we don't need and don't want to duplicate functionality that is provided by JDK without a good reason. Java will use own thread pool based implementation on systems that don't have asynchronous reads supported by JDK, and optimized implementations when such functionality is available.

forchid commented:
@katzyn Actually FilePathRetryOnInterrupt  can't prevent the session interrupted thread from closing the mvstore , and the interrupted mvstore can be corrupted. It's very dangerous.

katzyn commented:
If you want to fix it, try to do it properly. Java has everything you need for asynchronous IO.

grandinj commented:
@little-pan Please don't feel that you should give up, there are some decent ideas here.

I am not opposed to merging this as it currently stands because it adds a new backend that people can opt into if they want.

The only part that I don't like is re-routing of an existing backend.

@katzyn sure it would be better if it used java's asynch IO, but this is a decent attempt, and as it is a separate filestore backend, it doesn't affect any of the other backends
